### PR TITLE
JSONs gender

### PR DESCRIPTION
**Pull request by JaskRendix**
Genders for npcs and monsters.
This PR is for creating the data structure JSONs.

My question is: do we need to use the same field "gender" for both? Is it ok?

If not we could use "gender" for monsters and "identity" for NPCs as well as "genus".

If someone notices a wrong gender, then feel free to signal.

Some interesting numbers:
**create_npc** generates **213 NPCs with a JSON** and **12 NPCs without a JSON** (old maps) and there are _23 NPCs with a JSON that have never been created in a script_.

12: cfanatic1, cfanatic2, danteinstore, npc_aeble, npc_allie, npc_liela, npc_maples_mom, Rival_Dad, shopkeeperpapertown, Student1, Student2, Tuxemon_Collector

23: karrianna, penguin (!), professor-cardiling, professor-rockitten, professor-tweesher, red (npc_red), spyder_dragonscave_daeny, spyder_dryadsgrove_ferris, spyder_magician (!), spyder_nimrod_mace, spyder_professor (!), spyder_riverboatcaptainpapertown (!), spyder_wayfarer1_bismuth, spyder_wayfarer1_ratcher, xerogrund2, xerogrunt10, xerogrunt3, xerogrunt4, xerogrunt5, xerogrunt6, xerogrunt7, xerogrunt8, xerogrunt9
(!) no translation, only JSON

### CODE REVIEW

**Comment by kerizane** [comment]
Add "gender" to pydantic validation for NPCs.
Can we add the h validation to npc.monster,  but not to monster?

Could be done here or in future ticket:
Pick more appropriate genders for the npc monsters.
The monsters JSON should include "possible_genders", this should be a list of strings: male, female, neuter.
When you create monster you should pick one of the possible_genders at random. Most monsters can be male or female, some can only be neuter or only male or only female.

**Comment by JaskRendix** [comment]
> Add "gender" to pydantic validation for NPCs.

Done.

> Can we add the h validation to npc.monster, but not to monster?

PartyMemberModel?

>  Pick more appropriate genders for the npc monsters.

Work in progress

> The monsters JSON should include "possible_genders", this should be a list of strings: male, female, neuter.

I guess you're still referring to the PartyMemberModel?

**Comment by JaskRendix** [comment]
It would become like this? I need to know if we need to maintain gender and gender, or use another name.
```
{
  "slug": "spyder_citypark_edith",
  "gender": "female",
  [...]
  "monsters": [
    {
      "name": "Squabbit",
      [...]
      "possible_genders": "neutral"
    }
  ],
  "combat_front": "picnicker.png",
  "combat_back": "red_back.png"
}
```

**Comment by kerizane** [comment]
Gender should be on npc and PartyMemberModel (the monster an npc has) possible_genders should be on monster (because this is a property of the monster species).

**Comment by kerizane** [comment]
```
{
  "slug": "spyder_citypark_edith",
  "gender": "female",
  [...]
  "monsters": [
    {
      "name": "Squabbit",
      [...]
      "gender": "neutral"
    }
  ],
  "combat_front": "picnicker.png",
  "combat_back": "red_back.png"
}
```

**Comment by JaskRendix** [comment]
Ok, it was a bunch of stuff, but another step has been done. Let me know if you like the enum idea.

**Comment by JaskRendix** [comment]
Ah yeah, you said neuter, not neutral, I'll update after you check.

>  Pick more appropriate genders for the npc monsters.

Solved.

File: tuxemon/db.py [hunk 1]
```diff
@@ -227,6 +233,7 @@ class MonsterModel(BaseModel):
     shape: str = Field("", description="The shape of the monster")
     types: Sequence[str] = Field([], description="The type(s) of this monster")
     catch_rate: float = Field(0, description="The catch rate of the monster")
+    possible_gender: GenderType = Field(..., description="The gender of the monster")
```
Comment (kerizane) [inline tuxemon/db.py]:
```
possible_genders: Sequence[GenderType]
```

File: tuxemon/db.py [hunk 2]
```diff
@@ -398,6 +405,7 @@ class PartyMemberModel(BaseModel):
         ..., description="Modifier for experience this monster gives"
     )
     exp_req_mod: float = Field(..., description="Experience required modifier")
+    gender: str = Field(..., description="Gender of the monster")
```
Comment (kerizane) [inline tuxemon/db.py]: gender: GenderType

File: tuxemon/db.py [hunk 3]
```diff
@@ -408,6 +416,7 @@ def monster_exists(cls, v):
 
 class NpcModel(BaseModel):
     slug: str = Field(..., description="Slug of the name of the NPC")
+    gender: str = Field(..., description="Gender of the NPC")
```
Comment (kerizane) [inline tuxemon/db.py]: gender: GenderType

File: tuxemon/monster.py [hunk 4]
```diff
@@ -340,6 +341,7 @@ def load_from_db(self, slug: str) -> None:
         self.txmn_id = results.txmn_id
         self.height = results.height
         self.weight = results.weight
+        self.possible_gender = self.shape = GenderType.neutral
```
Comment (kerizane) [inline tuxemon/monster.py]: ??

Comment (JaskRendix) [reply tuxemon/monster.py]: sorry, confusion, I'll correct

**Comment by JaskRendix** [comment]
I'm going to fix neutral -> neuter.

File: tuxemon/db.py [hunk 5]
```diff
@@ -244,6 +250,7 @@ class MonsterModel(BaseModel):
     shape: MonsterShape = Field(..., description="The shape of the monster")
     types: Sequence[str] = Field([], description="The type(s) of this monster")
     catch_rate: float = Field(0, description="The catch rate of the monster")
+    possible_gender: Sequence[GenderType] = Field(..., description="The gender of the monster")
```
Comment (kerizane) [inline tuxemon/db.py]: I still think this should be plural: possible_genders also the description 

`description="Valid genders for the monster"`

**Comment by JaskRendix** [comment]
```
2022-09-25 16:08:30,936 - tuxemon.db - ERROR - validation failed for 'grinflare': 1 validation error for MonsterModel
possible_gender
  field required (type=value_error.missing)
```

Fixed neutral -> neuter, I guess we need to add **possible_gender** to the monster jsons too?

Something like:
possible_gender: Sequence[GenderType] = Field([], description="The gender of the monster")
and then
possible_gender: []

**Comment by kerizane** [comment]
Yeah, this will be a bit of work. To do this properly you will need to go through every monster and define what genders are possible.

Most will be ["male", "female"].
There are some with the male/female symbol in their name, so these will be ["male"] or ["female"].
And some might need to be ["neuter"].

If this will require too much work, or will need too much discussion, we can split it off into another ticket.

**Comment by JaskRendix** [comment]
No problem, I have already done the list of exceptions, because of the npc_monsters.
 
- only female: nudiflot_female, nudimind (evolution of nudiflot_female), viviteel (vivipere evolves into viviteel if female)
- only male: nudiflot_male, nudikill (evolution of nudiflot_male)
- only neuter: botbot, av8r, picc, mrmoswitch, k9, mk01_alpha, mk01_beta, birb-robo, chrome-robo, dark_robo, xeon, xeon_2, r0ck1tt3n, l3gk0, f7u1t3ra, d0llf1n, pythwire, ouroboutlet, sockeserp

Curiosity: is there a male evolution for vivipere? Yeah, but it's not added at the moment.

Think about this: we have 185 monsters, the NPCs use only 106!

**Review by kerizane — CHANGES_REQUESTED** [review]

**Comment by kerizane** [comment]
Neuter not fixed eithin

**Comment by JaskRendix** [comment]
Done as requested, renaming included.
Now I'm thinking that maybe I didn't understand the "neuter not fixed". Is it ok?

> When you create monster you should pick one of the possible_genders at random.

Just to inform that I have already something for this, but I'll open another PR.

File: tuxemon/monster.py [hunk 6]
```diff
@@ -227,6 +228,7 @@ def __init__(self, save_data: Optional[Mapping[str, Any]] = None) -> None:
         self.faint_cry = ""
         self.ai: Optional[ai.AI] = None
         self.owner: Optional[NPC] = None
+        self.possible_genders = GenderType
```
Comment (kerizane) [inline tuxemon/monster.py]:
```
self.possible_genders: List[GenderType] = []
```

**Review by kerizane — APPROVED** [review]

**Comment by kerizane** [comment]
Sorry, the following comment was meant to go on this PR:

"Have you tested combat against a wild monster and against an NPC?

If you run make validate from the base of the repo are there any errors?"

**Comment by kerizane** [comment]
If `make validate` doesn't complain, then it's fine.

**Comment by JaskRendix** [comment]
No errors. Make validate as before:
```
PYTHONPATH=. python scripts/schema.py --validate
pygame 2.1.2 (SDL 2.0.16, Python 3.10.6)
Hello from the pygame community. https://www.pygame.org/contribute.html
```

**Comment by Sanglorian** [comment]
> No problem, I have already done the list of exceptions, because of the npc_monsters.

Thanks for this list! I've added "Sex" as a variable in the wiki for monsters, and done it for all monsters except for those with the standard "male, female". I followed your list except for the glitch Tuxemon - which I think are fine to have sexes.

**Pull request merged**